### ISSUE TITLE
ACPI handler not activating for headphone/speaker events

### ISSUE THREAD

**Issue by nelsonni** [raphael/linux-samus#105]
Running version 4.4-5 on the latest Arch provides sound from speakers and headphone via PulseAudio in Gnome, but auto-switching between headphones and speakers is not working. This is a continuation of the issues outlined in #33 .

Examining the systemd journal shows that no events from `/etc/acpi/handler.sh` are being triggered on headphone jack removal/insertion. Found via:
```bash
journalctl | grep 'ACPI action'
```

**Comment by raphael** [comment]
Not sure why it's not journaled by ACPI events are triggered as witnessed by `acpi_listen`:
```
$ acpi_listen 
jack/headphone HEADPHONE plug
jack/headphone HEADPHONE unplug
```

**Comment by nelsonni** [comment]
A quick test with `acpi_listen` running in a bash terminal doesn't show any events occurring when headphones are plugged and unplugged from my system. Which makes me think that ACPI is doing it's job correctly, but no hardware event is being generated for it to respond to.

Attempting to find out whether the system is actually watching the pins for the headphone jack, I see that the following **asound** files exist:
```bash
$ ls -al /proc/asound/
total 0
dr-xr-xr-x   6 root root 0 Feb 12 14:50 .
dr-xr-xr-x 261 root root 0 Feb 12 10:03 ..
lrwxrwxrwx   1 root root 5 Feb 12 14:50 bdwrt5677 -> card0
dr-xr-xr-x   2 root root 0 Feb 12 14:50 card0
dr-xr-xr-x   2 root root 0 Feb 12 14:50 card1
-r--r--r--   1 root root 0 Feb 12 14:50 cards
-r--r--r--   1 root root 0 Feb 12 14:50 devices
lrwxrwxrwx   1 root root 5 Feb 12 14:50 HDMI -> card1
-r--r--r--   1 root root 0 Feb 12 14:50 hwdep
-r--r--r--   1 root root 0 Feb 12 14:50 modules
dr-xr-xr-x   2 root root 0 Feb 12 14:50 oss
-r--r--r--   1 root root 0 Feb 12 14:50 pcm
dr-xr-xr-x   2 root root 0 Feb 12 14:50 seq
-r--r--r--   1 root root 0 Feb 12 14:50 timers
-r--r--r--   1 root root 0 Feb 12 14:50 version
```
But ``card0`` only contains a **id** file, and ``card1/codec#0`` does not show any changes to ``Pin-ctls: 0x00:` when plugging/unplugging headphones. The ``card1/codec#0`` result isn't surprising however, since **bdwrt5677** is supposed to be driving audio and it's pointed at **card0**.

**Comment by nelsonni** [comment]
According to this article, the CONFIG_SND_HDA_INPUT_JACK kernel config option needs to be set in order for hardware events to be generated on the headphone jack: http://unix.stackexchange.com/a/128007

A quick ``zgrep JACK /proc/config.gz`` shows me the following:
```bash
$ zgrep JACK /proc/config.gz
CONFIG_SND_JACK=y
CONFIG_USB_SERIAL_CYBERJACK=m
# CONFIG_EXTCON_ADC_JACK is not set
```

**Comment by raphael** [comment]
strange, are you using `4.4.1-5ph` (`uname -r`) ? that's what I'm running for the test above. As far as I know `CONFIG_SND_HDA_INPUT_JACK` has been superseeded with `CONFIG_SND_JACK`.

**Comment by nelsonni** [comment]
```bash
$ uname -r
4.4.1-5ph
```
As far as ``CONFIG_SND_JACK`` superseeding ``CONFIG_SND_HDA_INPUT_JACK``, are you referring to **[PATCH] ALSA: hda: Make core jack usage unconditional** from the ALSA Project in 2012? http://comments.gmane.org/gmane.linux.alsa.devel/100531

**Comment by nelsonni** [comment]
Just for posterity, I'm also including a link to **MORE NOTES ON HD-AUDIO DRIVER** that Takashi Iwai added to _kernel.org_: https://www.kernel.org/doc/Documentation/sound/alsa/HD-Audio.txt

It appears that there were several changes necessitated by HD-Audio drivers within the kernel community.

**Comment by raphael** [comment]
I'm referring to this: http://article.gmane.org/gmane.linux.alsa.devel/137933/match=config_snd_hda_input_jack

Maybe try a different headphone? not sure what would cause ACPI to work on my system and not yours. Same kernel, same hardware...

**Comment by nelsonni** [comment]
I had the same results of ``acpi_listen`` returning no events with all of the following:
* Bose QuietComfort 15
* Yurbuds Signature Series ITX-3000
* Samsung earphones that came with Google Nexus S (2010)

I'm going to attempt to track back from the ``acpi_listen`` command to see whether events are being generated by the hardware and caught, but simply not passed along correctly. I'll post a patch if I figure it out.

**Comment by nelsonni** [comment]
I found the culprit, and a potential fix thanks to a blog post about Acer Chromebook 13 configurations by [Ronny Lorenz](https://www.tbi.univie.ac.at/~ronny/acer-cb5-311.html).

When `xplug` events are generated and captured by `acpi` for handling, and `PulseAudio` is enabled, the switching mechanism between `headphone` and `speaker` needs to be handled by `pacmd` instead of `alsactl`.

I've got a patch created that integrates these changes. Once switched, `acpi_listen` has the following output upon plugging in headphones on my system:
```bash
$ acpi_listen
jack/headphone HEADPHONE plug
jack/microphone MICROPHONE plug
```
